anaconda_packaging_utils/api/repodata_api.py

```diff
@@ -0,0 +1,319 @@
+    OSX_X86_32 = "osx-32"
+    WIN_64 = "win-64"
+    WIN_32 = "win-32"
+    NO_ARCH = "noarch"
+
+
+_BASE_REPODATA_URL: Final[str] = "https://repo.anaconda.com/pkgs"
+
+
+class ApiException(BaseApiException):
+    """
+    Indicates an exception occurred with this API.
+    """
+
+    pass
+
+
+# Schema for the `channeldata.json` structure. As of writing, this only validates two fields of the data structure, as
+# we only use this JSON file to determine which architectures are supported per channel.
+_CHANNELDATA_JSON_SCHEMA: Final[SchemaType] = {
+    "type": "object",
+    "required": ["channeldata_version", "subdirs"],
+    "properties": {
+        "channeldata_version": {"type": "integer"},
+        "subdirs": {"type": "array", "items": {"type": "string"}},
+    },
+}
+
+
+@dataclass
+class RepodataMetadata:
+    """
+    Metadata section in a `repodata.json` blob
+    """
+
+    ## Required fields ##
+    subdir: str
+    ## Optional fields ##
+    arch: Optional[str] = None
+    platform: Optional[str] = None
+
+
+@dataclass
+class PackageData:
+    """
+    Per-package data stored in a `repodata.json` blob
+    """
+
+    ## Required fields ##
+    build: str
+    build_number: int
+    depends: list[str]
+    md5: str
+    sha256: str
+    name: str
+    size: int
+    version: str
+    subdir: str
+    ## Optional fields ##
+    timestamp: Optional[int] = None
+    date: Optional[str] = None
+    track_features: Optional[str] = None
+    license: Optional[str] = None
+    license_family: Optional[str] = None
+
+    # TODO rm: Enforce type checking when this PR is released
+    #   https://github.com/conda/conda/pull/13385
+    @no_type_check
+    def __lt__(self, other: object) -> bool:
+        """
+        Allows for version comparisons between two `PackageData` instances
+        NOTE: We are relying on the default implementation of `__eq__()` for dataclasses, which checks that all fields
+              are equivalent.
+        NOTE: Type checking is disabled for this function as `VersionOrder`
+        :param other: Object to compare against this one.
+        :returns: If this instance has a lower version than the compared instance.
+        """
+        if not isinstance(other, PackageData):
+            return False
+        # Comparing versions between two different software projects/packages is meaningless.
+        if self.name != other.name:
+            return False
+        try:
+            self_v = VersionOrder(self.version)
+            other_v = VersionOrder(other.version)
+            # If two versions are equal, fallback to comparing the build number
+            if self_v == other_v:
+                return self.build_number < other.build_number
+            return self_v < other_v
+        except InvalidVersionSpec:
+            # If a `VersionOrder` object failed to be created, we are not able to accurately compare these objects.
+            pass
+        return False
+
+
+@dataclass
+class Repodata:
```

Comment:
So this seems to be something similar to the Repodata class in anaconda-channel-scanner. The only difference is that there are more checks in this implementation, and that it processes all the info. Anaconda-channel-scanner just reads what it needs for the dependency analysis and also structures the package data into  a few levels: package-name/python-version/package-version/build-number.

Reply:
Good to know. For Mark's needs (and I think percy's as well), all the data needs to be available at once. I wanted to create some more intelligent caching setup, but it also sounds like Cheng is on the warpath to replace these unscalable JSON files with something else.

So this is more of a stop-gap until we know what the "new and improved" version will be.